### PR TITLE
Bypass v9

### PR DESCRIPTION
**Pull request by regit**
Update of #2278.

Callback in capture method now return a status which allows to detect the bypass failure in lower level. A rebase has been done and flow flags had to be  updated to get some free room.

PR builds:
- PR regit: https://buildbot.openinfosecfoundation.org/builders/regit/builds/199
- PR regit-pcap: https://buildbot.openinfosecfoundation.org/builders/regit-pcap/builds/195

### CODE REVIEW

File: src/detect-filestore.c [hunk 1]
```diff
@@ -408,3 +416,39 @@ static void DetectFilestoreFree(void *ptr)
...
+
+    de_ctx->flags |= DE_QUIET;
+
+    de_ctx->sig_list = SigInit(de_ctx,"alert http any any -> any any "
+                               "(bypass; filestore; "
+                               "content:\"message\"; http_host; "
+                               "sid:1;)");
+    FAIL_IF_NOT_NULL(de_ctx->sig_list);
+
+    if (de_ctx != NULL)
```
Comment (inliniac) [inline src/detect-filestore.c]: de_ctx can't be NULL here

File: src/flow-manager.c [hunk 2]
```diff
@@ -264,6 +264,15 @@ static int FlowManagerFlowTimedOut(Flow *f, struct timeval *ts)
     }
 
     int server = 0, client = 0;
+
+    int state = SC_ATOMIC_GET(f->flow_state);
+    if (state == FLOW_STATE_BYPASSED)  {
+        if (FlowForceReassemblyNeedReassembly(f, &server, &client) == 1) {
```
Comment (inliniac) [inline src/flow-manager.c]: I don't see why this commit is needed. The flow will be handled by the flow timeout mechanism rather soon after the bypass, as it's timeout is now really low.

Comment (inliniac) [inline src/flow-manager.c]: Commit message is hard to follow.

File: src/flow.c [hunk 3]
```diff
@@ -225,28 +225,26 @@ void FlowHandlePacketUpdate(Flow *f, Packet *p)
...
-                f->flags |= FLOW_TO_SRC_SEEN;
                 p->flowflags |= FLOW_PKT_TOCLIENT_FIRST;
             }
         }
+        f->tosrcpktcnt++;
+        f->tosrcbytecnt += GET_PKT_LEN(p);
     }
 
-    if ((f->flags & (FLOW_TO_DST_SEEN|FLOW_TO_SRC_SEEN)) == (FLOW_TO_DST_SEEN|FLOW_TO_SRC_SEEN)) {
+    if (f->tosrcpktcnt && f->todstpktcnt) {
```
Comment (inliniac) [inline src/flow.c]: so the point of the flags logic was that we don't set 'established' on icmp error packets coming in. I think we lost that here.

File: src/flow-manager.c (hunk 2 above)
Comment (regit) [inline src/flow-manager.c]: I did not get log after  30 sec. flow entry were only shown when shunting down suricata.

**Comment by inliniac** [comment]
So I think the FLOW_BYPASS_FAILED flag needs some more thought. If we assume that the bypass can fail for multiple reasons:
1. capture method not supporting it
2. capture method not setup properly (e.g. no NFQ rule that actually enforces it)
3. capture method with limits (e.g. max 1000 nic hw filters)

Then we have to handle failure gracefully. I think we should unconditionally flag the flow as 'bypassed' and stop processing packets as early as possible. I'm thinking we exit 'FlowWorker' right after it looked up the flow and checked that this 'bypassed' flag is set.

Such a generic flag would replace the 'bypass failed' flag I think.

**Comment by inliniac** [comment]
Another thing is that if we still continue to see packets the very short flow timeout connect to the bypass state may actually be harmful. If the flow is discard from suricata but more packets come in, we'll just create a new flow for that traffic.

File: src/flow-manager.c (hunk 2 above)
Comment (inliniac) [inline src/flow-manager.c]: That is strange. Looks like this fix is hiding another issue then.

**Comment by regit** [comment]
Yes on timeout, even if 5 seconds is long here. But if we have a flag that discard all new packets for the flow then that should be not anymore a potential problem.

**Comment by inliniac** [comment]
Not sure I follow. If we have a 5 sec timeout, the flow is gone after 5 sec. What if a packet for that flow comes in after 10 sec? A new flow will be created. A ssh session would be an example of this. Of course we have the same issue with our regular timeouts, but they are set quite a bit higher by default.

**Comment by regit** [comment]
If we omit Netfilter misconfiguration, the bypass call is acknowledged by the lower level so we should not receive any new packets from the flow after the call. Only already queued packets will be seen by suricata and a 5 seconds delay between 2 packets of the same flow is really unlikely to occur.
I fully agree on the early bypass comment, it seems a really good idea to exit as soon as possible and thus provide a "software" bypass fallback.

**Comment by inliniac** [comment]
Another case may be that the capture method sets a timeout on the filter it uses, and may discard the filter for a flow too early. Or maybe it discards a filter because it needs to make space for new filters. In general I think the bypass should be handled completely inside suricata, and then we offload to capture if we're lucky.

**Comment by regit** [comment]
We have a timing issue to consider if we do that: suppose the HW bypass method is bypassing for a while a flow then entry for the flow is removed. In that case we could see packets from a flow after a long time (without knowing if it was alive or not). So if we want to deal with that it means we should keep a bypassed flow forever or assume we can see an already bypassed flow as new if ever HW is playing a trick to us.

**Comment by inliniac** [comment]
Yeah this is a trade off. Perhaps it would make sense to keep a 'light' flow around for a longer time. Just the flow, but all applayer/flowbit/etc stuff already freed.

**Comment by regit** [comment]
So to summarize, I think we should implement a suricata-only bypass that get the bypassed packets out of flow worker just after the lookup.
But in case we have a lower level method that is doing the bypass, we may have a mixed timeout/lower level bypass stop that would cause some problem and make it almost useless. So I think the suricata-only bypass should be or disabled or should have low timeout value when used with a lower level bypass.

**Comment by regit** [comment]
One other important point on callback failure. It could also be due to the fact that the protocol is not supported by callback method. For instance, AF_PACKET is not supporting now IPv6 so it was convenient to be able to say for the capture method to say, well I can't bypass. So it may really be interesting to make the distinction between failed bypassed and successful bypass.

File: src/detect-filestore.c (hunk 1 above)
Comment (regit) [inline src/detect-filestore.c]: ack

**Comment by inliniac** [comment]
On 26-09-16 11:19, Eric Leblond wrote:
> *@regit* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In src/flow-manager.c <https://github.com/inliniac/suricata/pull/2282>:
> 
>> @@ -264,6 +264,15 @@ static int FlowManagerFlowTimedOut(Flow *f, struct timeval *ts)
>      }
>  
>      int server = 0, client = 0;
> +
> +    int state = SC_ATOMIC_GET(f->flow_state);
> +    if (state == FLOW_STATE_BYPASSED)  {
> +        if (FlowForceReassemblyNeedReassembly(f, &server, &client) == 1) {
> 
> I did not see that before a rebase on latest code. Could it be related
> to recent flow manager changes ?

It shouldn't but you never know :)

**Comment by regit** [comment]
I propose to introduce two state in flow_state:

1. LOCAL_BYPASSED that is set when bypass callback is failing
2. CAPTURE_BYPASSED that is is set when bypass callback is a success

LOCAL_BYPASSED will trigger the same timeout policy as a ESTABLISHED and suricata will get packets out of flow worker after the hash lookup.

For CAPTURE_BYPASSED, we keep the same timeout policy as the one which is in implemented by this PR. 

This means we can get rid of the BYPASS FAILED which is replace by LOCAL_BYPASSED.

**Comment by inliniac** [comment]
If LOCAL_BYPASSED is equal to ESTABLISHED, how to do transition to 'CLOSED'?

**Comment by regit** [comment]
Right. We could have a reasonnably short timeout for instance inherited from emergency ?

**Comment by regit** [comment]
Note: we discard the packets for the flow and don't do any tracking so we can't relay on closing info RST, FIN because we don't track TCP window and ack.

**Comment by inliniac** [comment]
Might make sense to stay close to 'closed', but otoh lots of ppl set that to 0. New setting with a default of 30/60sec or so? Not very pretty.

**Comment by regit** [comment]
I was thinking about using emergency established because it is a similar case: flow is there and we don't want to spend too much time with it.

**Comment by inliniac** [comment]
I think in many cases it's set too low to be useful.

**Comment by inliniac** [comment]
A key difference with regular flow handling is that performance here may very well depend on *long* timeouts. The shorter the timeout the higher the change of a new mid-stream flow based on bypassed traffic.

**Comment by regit** [comment]
Well, if a bypassed flow don't send anything for 100 sec, actual default then we don't really care about it in most cases.

**Comment by regit** [comment]
I've seen that it is not trivial to access the emergency timeout anymore. We could consider that the feature is important enough for users for them to have new timeout to set.

**Comment by inliniac** [comment]
Yeah I think that makes sense.

**Comment by regit** [comment]
ok, we will add a new variable for local bypass flow timeout.
